File: backend/app/crud/audit.py
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.audit_log import AuditLog


def create(db: Session, user_id: int | None, action: str, entity_type: str, entity_id: int | None, details: dict) -> None:
    db.add(
        AuditLog(user_id=user_id, action=action, entity_type=entity_type, entity_id=entity_id, details=details)
    )
    db.commit()


def list_paginated(
    db: Session, page: int = 1, page_size: int = 50, entity_type: str | None = None
) -> tuple[list[AuditLog], int]:
    stmt = select(AuditLog)
    if entity_type:
        stmt = stmt.where(AuditLog.entity_type == entity_type)

    total = db.scalar(select(func.count()).select_from(stmt.subquery())) or 0
    stmt = stmt.order_by(AuditLog.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
    return list(db.scalars(stmt)), total

File: workers/src/repo/audit.ts
import { nowIso } from "../lib/time";

export interface AuditLogRow {
  id: number;
  user_id: number | null;
  email: string | null;
  action: string;
  entity_type: string;
  entity_id: number | null;
  details: string | null;
  created_at: string;
}

/** `details` is a JSON document stored as TEXT, matching the SQLAlchemy JSON column. */
export async function create(
  db: D1Database,
  userId: number | null,
  action: string,
  entityType: string,
  entityId: number | null,
  details: Record<string, unknown>
): Promise<void> {
  await db
    .prepare(
      `INSERT INTO audit_logs (user_id, action, entity_type, entity_id, details, created_at)
       VALUES (?, ?, ?, ?, ?, ?)`
    )
    .bind(userId, action, entityType, entityId, JSON.stringify(details), nowIso())
    .run();
}

export async function listPaginated(
  db: D1Database,
  page: number,
  pageSize: number,
  entityType?: string
): Promise<{ items: AuditLogRow[]; total: number }> {
  const clause = entityType ? " WHERE a.entity_type = ?" : "";
  const params: unknown[] = entityType ? [entityType] : [];

  const countRow = await db
    .prepare(`SELECT COUNT(*) AS count FROM audit_logs a${clause}`)
    .bind(...params)
    .first<{ count: number }>();

  const { results } = await db
    .prepare(
      `SELECT a.*, u.email AS email FROM audit_logs a
       LEFT JOIN users u ON u.id = a.user_id${clause}
       ORDER BY a.created_at DESC LIMIT ? OFFSET ?`
    )
    .bind(...params, pageSize, (page - 1) * pageSize)
    .all<AuditLogRow>();

  return { items: results, total: countRow?.count ?? 0 };
}
